fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -167,6 +186,16 @@ def _find_rank(self, param: Parameter) -> Tuple[OSS, int]:
         assert False, "This parameter is not present in an optimizer, this should not happen"
         return (None, -1)
 
+    def _gatekeeper(self) -> None:
+        # Make sure that all the asynchronous calls have concluded before moving on. Consume the futures
+        # and execute the delayed actions (release gradients, unroll the buckets)
+        for optimizer in self.sharded_optimizers:
+            # Empty in-flight futures
+            Variable._execution_engine.queue_callback(optimizer._consume_work_handles)
```

Comment:
Why do you queue callbacks here? The backward should already be finished for the part that you are interested in and so you should be able to call these functions directly no?

Reply:
just saw that now (after my other reply), sorry for the async. Yes I think you're right, it's some copy/pasta of an old codebase, it does not seem to make sense in that case, I agree.